File: solutions/[147]对链表进行插入排序.cpp
#include <cstdlib>
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode* insertionSortList(ListNode* head) {
		ListNode* dummy = new ListNode(-1);
		dummy->next = head;
		ListNode* cur = head;
		ListNode* tail = head;
		while (cur != nullptr) {
			ListNode* next_cur = cur->next;
			ListNode* tmp = dummy;
			for (; tmp->next->val < cur->val && tmp->next != cur; tmp = tmp->next);
			if (tmp->next != cur) {
				ListNode* tmp_next = tmp->next;
				tmp->next = cur;
				tail->next = cur->next;
				cur->next = tmp_next;
			}
			else {
				tail = cur;
			}
			cur = next_cur;
		}
		return dummy->next;
	}
};

File: solutions/[147]对链表进行插入排序.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        if head is None:
            return head

        dummy = ListNode(float('-inf'))
        dummy.next = head
        last_sorted = head
        p = head.next

        while p is not None:
            if last_sorted.val <= p.val:
                last_sorted = last_sorted.next
            else:
                prev = dummy
                while prev.next.val <= p.val:
                    prev = prev.next
                last_sorted.next = p.next
                p.next = prev.next
                prev.next = p
            p = last_sorted.next
        
        return dummy.next
